accepted/2020/binary-data/binary-data.md

```diff
@@ -0,0 +1,177 @@
+    - String encoding is assumed to be UTF8
+    - Object serialization is done with `System.Text.Json`, with specific options
+
+### Non-Goals
+
+* Allocation free zero-copying semantics. The primary goal is convenience and
+  usability, not perf. Performance critical scenarios should be enabled by
+  providing overloads with the existing low-level concepts, such as span.
+
+## Design
+
+We're concerned with hard coding the type to `System.Text.Json`, but we
+acknowledge that defaults are very valuable. If we need to support other
+serializers we can provide additional arguments or APIs.
+
+This type can't live in corlib because it would depend on `System.Text.Json`
+(and whatever future serializers it needs to support). Hence, most areas in the
+BCL wouldn't be able to take `BinaryData`, but technologies above the BCL could
+(Microsoft.Extensions, ASP.NET Core, Azure SDK etc). While that might be
+unfortunate, we believe it's acceptable because most of the usability gains
+would be in higher levels anyway.
+
+Thus, we shouldn't think of this as an exchange type in the BCL. We already have
+exchange types; the point of this type is unify them by providing conversions;
+the point isn't to provide yet another type. Thus, we don't see the need to, for
+example, provide new virtual methods on Stream.
+
+We have two options to design the type:
+
+1. We could decide to model the construction via JSON serialization as a factory
+   method. This ensures we can easily support additional serialization formats
+   as necessary and remain linker friendly.
+
+2. But we need to balance implementation flexibility concerns with usability.
+   The primary goal of this type is to provide usability via conveniences. If we
+   don't make the type super usable, the feature will fail. If we feel that the
+   factories aren't discoverable enough, we should combine it with constructors,
+   even if that reduces our flexibility.
+
+### Shipping vehicle
+
+**Assembly**: System.BinaryData.dll
+**NuGet Package**: System.BinaryData (stable)
+
+### API
+
+```C#
+namespace System
+{
+    public sealed class BinaryData
+    {
+        public static BinaryData FromBytes(ReadOnlyMemory<byte> data);
+        public static BinaryData FromBytes(ReadOnlySpan<byte> data);
+        public static BinaryData FromBytes(byte[] data);
+        public static BinaryData FromObjectAsJson<T>(T jsonSerializable, CancellationToken cancellationToken = default);
+        public static Task<BinaryData> FromObjectAsJsonAsync<T>(T jsonSerializable, CancellationToken cancellationToken = default);
+        public static BinaryData FromStream(Stream stream);
+        public static Task<BinaryData> FromStreamAsync(Stream stream, CancellationToken cancellationToken = default);
+        public static BinaryData FromString(string data);
+        public static implicit operator ReadOnlyMemory<byte>(BinaryData data);
+        public ReadOnlyMemory<byte> ToBytes();
```

Comment:
Please make this `AsBytes` since it is non-allocating. It is a C# community convention. It is in framework APIs too. https://stackoverflow.com/a/13867955/294804

Reply:
That's a good point, we missed that. @KrzysztofCwalina @bartonjs, I assume we agree with that?

Reply:
`AsBytes` does seem more idiomatic than `ToBytes`.  ImmutableArray mixes an As(Memory/Span) and a To(Builder), so it's not unheard of for a single type to use both.

Reply:
I think this was considered but we felt it wasn't necessarily worth deviating from the To/From pattern used throughout the type.

Reply:
For F# having the implicit conversions available as normal methods is helpful, quite painful having to invoke the op_Implicit method.

Reply:
I think we've done this after the UX study to help customers find all related `To*` overloads because they are grouped together.

Reply:
OK, based on conversations and too keep things consistent, I've spec'ed it as `AsBytes()`. @JoshLove-msft and @pakrym please tell me if you feel strongly that this is the wrong thing.

Reply:
I honestly think the distinction between As and To is something only a few understand and for the rest it will be just an annoyance, as users will have to continuously think "is the conversion I want a To or an As conversion". When everything starts with To, you just type "To" and select from intellisense list.

Reply:
It is a common convention for those using LINQ or Entity Framework regularly. Personally, I understood it within the first 2 years of writing C# professionally. It is necessary to distinguish between `As` and `To` because in LINQ, `To` resolves the query while `As` does not. For this use case, our usage of `BinaryData` looks bizarre currently as it is littered with `ToBytes().ToArray()`, which infers we're allocating a new object to then allocate that as an array, which is not the case. Intellisense shouldn't trump common patterns that are reinforced by frameworks and community code, IMO.

Reply:
I feel myself agreeing with @MiYanni. @bartonjs?

Reply:
Yeah, I'm with the "ToBytes().ToArray() looks 'two' expensive", so my preference is for `AsBytes` (or AsMemory so we can also easily add AsSpan)

Reply:
To in LINQ is very expensive in comparison to As. Also, LINQ's As allocates. In other words, the As convention does not mean "no allocation" and the To convention does not mean allocation either. For example Memory<char>.ToString might not actually allocate a string.

Reply:
In our UX study for this type, we initially had a `Bytes` property instead of `ToBytes()` and participants were struggling.  One of them even turned it into a `Stream` in an attempt to get a byte array out of it:

> "I looked at the ToStream directly because it looked like something natural to me, probably missed some properties on the way. Maybe it will be more intuitive if there were a method ToBytes() instead of Bytes property."

We switched to `ToBytes()` half way through and no one else struggled.  Having all the static `From*()` and instance `To*()` options align in intellisense makes it very easy for customers to discover all their options for transforming binary data.  We'd see customers scrolling back and forth over just the `From*()`s and `To*()`s in intellisense when attempting each task.  Moving one of the options to the top of the list with `AsBytes()` would make that worse.

Reply:
```C#
public ReadOnlyMemory<byte> AsMemory();
public ReadOnlySpan<byte> AsSpan();
public byte[] ToBytes();
```

?

Reply:
@bartonjs, I think the reasons we have As and To in these APIs are:

1. The API is low level API focusing on perf. We wanted to be extra explicit about the difference in perf characteristics.
2. The difference in perf characteristic between AsSpan and ToBytes is huge.

In the case of BinaryData the perf difference between ToStream and AsBytes is pretty much irrelevant, and the type is focusing on usability, not on writing low level perf to the metal code.